components/homme/src/theta-l_kokkos/cxx/cxx_f90_interface_theta.cpp

```diff
@@ -345,12 +347,14 @@ void init_functors_c (const bool& allocate_buffer)
 #ifdef HOMME_ENABLE_COMPOSE
   else                           c.create_if_not_there<ComposeTransport>();
 #endif
-  auto& hvf  = c.create_if_not_there<HyperviscosityFunctor>();
-  auto& ff   = c.create_if_not_there<ForcingFunctor>();
-  auto& diag = c.create_if_not_there<Diagnostics> (elems.num_elems(),params.theta_hydrostatic_mode);
-  auto& vrm  = c.create_if_not_there<VerticalRemapManager>(elems.num_elems());
+  auto& hvf     = c.create_if_not_there<HyperviscosityFunctor>();
+  auto& ff      = c.create_if_not_there<ForcingFunctor>();
+  auto& diag    = c.create_if_not_there<Diagnostics> (elems.num_elems(),params.theta_hydrostatic_mode);
+  auto& vrm     = c.create_if_not_there<VerticalRemapManager>(elems.num_elems());
 
-  auto& fbm  = c.create_if_not_there<FunctorsBuffersManager>();
+  auto& fbm     = c.create_if_not_there<FunctorsBuffersManager>();
+
+//OG why are if-statement here -- above calls define which constructor is called
```

Comment:
I don't recall exactly the details, but I think it was related with the order in which things are inited when running eamxx. IIRC, in eamxx we create the functors before the Elements structures are ready (we still need to receive the pointers from the atm driver), so we create them "bare", but with enough info to establish how much memory they need from the FunctorsBufferManager, so that the atm driver can create the scratch buffers. Basically this is the sequence from the AD perspective

1. create atm processes (include dynamics)
  a. dyn creates all functors, does not set elements/tracers views
  b. dyn computes needs for buffers
2. get buffers requirements from all atm processes
3. allocate atm fields
4. pass fields and buffer to atm processes
  a. dyn can set ptrs inside Homme elements/tracers views, and complete functors setup

@tcclevenger might remember more details though.

Reply:
That correct, we basically needed to create them inside EAMxx, but without actually having all the information to do the setup.

Each of these classes has a constructor which defines the base variables needed to talk to the buffer manager (I think just `num_elem` and mabye `params`) and then stores a bool which says "this wasn't actually setup". So here, if that is the case, we finish setting up the class.

On the other hand, if the class was constructed fully, then the class store the bool as "already setup" and nothing happens below.